Removing Authorization header on cross host redirection

Satheesh-Balachandran opened the pull request.

Satheesh-Balachandran commented:
Related Issue: #282
@eyalbe4 Please review

Satheesh-Balachandran commented:
> I recommend using the code in [https://github.com/jfrog/build-info project](https://github.com/jfrog/build-info%C2%A0project) as a reference for this implementation. All of the following comments are implemented in the build-info project. You can search for the ORIGINAL_HOST_CONTEXT_PARAM const in the build-info code, for quick reference.
> 
> 1. The original host is saved in the context. It is done however for the HTTP GET only. Let's create a PreemptiveRedirectStrategy class.
> 2. Let's use a constant for the original host (ORIGINAL_HOST_CONTEXT_PARAM is a good choice).
> 3. There's probably no need to remove the Authorization after it is set. We can first check we need to add it, and we don't, skip adding it. The easiest thing would probably be copying the build-info code.

@eyalbe4 Made the changes. Used the same interceptor because it is more closely related to the Auth and we already have one.

eyalbe4 commented on httpClient/src/main/java/org/jfrog/artifactory/client/httpClient/http/auth/PreemptiveAuthInterceptor.java:
```diff
@@ -20,20 +24,51 @@
+
     @Override
     public void process(HttpRequest request, HttpContext context) throws HttpException, IOException {
+
```
Redundant new-line here.

eyalbe4 commented on httpClient/src/main/java/org/jfrog/artifactory/client/httpClient/http/auth/PreemptiveAuthInterceptor.java:
```diff
@@ -20,20 +24,51 @@
         }
     }
+
+    private boolean shouldSetAuthScheme(final HttpRequest request, final HttpContext context) {
```
The code of this method is not indented correctly.

Satheesh-Balachandran replied:
Fixed

eyalbe4 commented on services/src/main/groovy/org/jfrog/artifactory/client/impl/ArtifactoryImpl.java:
```diff
@@ -345,12 +336,22 @@ public String delete(String path) throws IOException {
+
+    public HttpResponse execute(HttpUriRequest request) throws IOException {
+        HttpClientContext clientContext = HttpClientContext.create();
+        if(clientContext.getAttribute(ORIGINAL_HOST_CONTEXT_PARAM) == null) {
```
Missing white-space after **if**.

Satheesh-Balachandran replied:
Fixed

Satheesh-Balachandran commented:
@eyalbe4 Fixed the test case failures and java formatting.

eyalbe4 commented on services/src/test/java/org/jfrog/artifactory/client/ArtifactoryTests.java:
```diff
@@ -130,28 +134,22 @@ public void socketTimeoutBuilderTest() {
     @Test
     public void addInterceptorTest() {
```
There seems be an issue with the new logic applied to this test. You are checking the value of the the *interceptor1Visits* variable before the interceptor had a chance to increment it, so the this test actually tests nothing right now.
The reason why the original code of the test (which seems to good) failed, is because a real bug, which was introduced by your code. The tests failed because this line **builder.build().restCall(req);** now throws an IOException, so the interceptors the test defines are never invoked. The reason why the restCall fails is because you're no longer allowing unauthenticated requests, which is s serious compatibility breakage.
The test code can be actually improved, so that it also fails if the restCall fails, with a clear error message.

eyalbe4 commented on httpClient/src/main/java/org/jfrog/artifactory/client/httpClient/http/auth/PreemptiveAuthInterceptor.java:
```diff
@@ -1,39 +1,72 @@
+				Credentials creds = credsProvider
+						.getCredentials(new AuthScope(targetHost.getHostName(), targetHost.getPort()));
+				if (creds == null) {
+					throw new HttpResponseException(HttpStatus.SC_UNAUTHORIZED, "No credentials for preemptive authentication");
```
You are no longer allowing unauthenticated requests here.
I suggest modifying the code as follows. Instead of:
```
				if (creds == null) {
					throw new HttpResponseException(HttpStatus.SC_UNAUTHORIZED, "No credentials for preemptive authentication");
				}
				BasicScheme authScheme = new BasicScheme();
				authState.update(authScheme, creds);
```
try doing:
```
				if (creds != null) {
					authState.update(new BasicScheme(), creds);
				}
``

Satheesh-Balachandran replied:
Correct. I was so wrong here overriding the original functionality of this file. Made the necessary changes. Thanks!

eyalbe4 commented on services/src/main/groovy/org/jfrog/artifactory/client/impl/ArtifactoryImpl.java:
```diff
@@ -28,329 +31,337 @@
-        }
-        return Util.responseToString(httpResponse);
-    }
+	private static final String ORIGINAL_HOST_CONTEXT_PARAM = "original.host.context.param";
```
Could it be that you used some kind a formatting tool on the whole file? This highly not recommended when you're contributing code, because it makes it hard to review the code. In addition, you're adding unnecessary changes to the code, which affects the history for whoever tries to read the code in the future.

Satheesh-Balachandran replied:
Totally Agree. But not sure why it didn't show up in my diff tool. Probably I need to switch to a better UI version of git diff tooling.

Satheesh-Balachandran commented:
@eyalbe4 Requesting for review again!. I fixed to use the code with better formatting and removed unwanted review challenges. Hope this will fix all the issues.

eyalbe4 commented on services/src/main/groovy/org/jfrog/artifactory/client/impl/ArtifactoryImpl.java:
```diff
@@ -28,6 +29,8 @@
  */
 public class ArtifactoryImpl implements Artifactory {
 
+	private static final String ORIGINAL_HOST_CONTEXT_PARAM = "original.host.context.param";
```
This constant needs to be defined as *public final*, so that it can be also used in the *PreemptiveAuthInterceptor* class. There's no need for the *PreemptiveAuthInterceptor* to also declare the same variable.
